bms/bms.1
.TH BMS 1 "" "MandelSpawn"
.SH NAME 
bms \- Batch MandelSpawn, a network Mandelbrot program.
.SH SYNOPSIS  
.B bms
[
.B \-options 
] ...
.SH DESCRIPTION  
.I Bms
is a program for making pictures of the Mandelbrot and Julia sets.
It is a batch-mode version of the interactive, X-based
.I xms
program (see
.IR xms (1))
for full information).
.PP
Bms writes the picture on its standard output in Portable Greymap
.I (pgm) 
or Portable Pixmap
.I (ppm)
format.  In the former case, the iteration counts are used directly as 
greyscale pixel values, so the interior of the Mandelbrot set will appear
white, not black.  The picture can be further manipulated using the
tools from Jef Poskanzer's PBMPLUS image processing and conversion
package.  A typical example is the pipeline

.PP
	bms -iterations 256 -width 600 -height 400 | 
.br
		pnminvert |
.br
		pgmtops -rle |
.br
		lpr -Pps
.PP
which can be used to print a black-on-white Mandelbrot image on a
PostScript printer.
.PP
.SH "COMMAND LINE OPTIONS"
.PP
Bms supports the following command line options:
.IP "\fB\-iterations\fP \fIn\fP"
Specifies the maximum number of iterations.  The default is 250.
.IP "\fB\-colour\fP"
Generate colour
.I (.ppm)
output instead of the default of black-and-white
.I (.pgm)
output.
.IP "\fB\-colours\fP \fIn\fP"
Specifies the number of colours, as in 
.IR xms(1) .
Use together with the
.B -colour
option.
.IP "\fB\-spectrum\fP \fIcolour\-colour\- ... \-colour\fP"
Specifies the colours to use, as in 
.IR xms(1) .
Use together with the
.B -colour
option.
.IP "\fB\-x\fP, \fB\-y\fP, \fB\-range\fP"
Specify the center coordinates and horizontal width, respectively, of
the part of the Mandelbrot or Julia set to be calculated.
.IP "\fB\-julia\fP"
Calculate a Julia set instead of the Mandelbrot
set.  The "\fB\-cx\fP" and "\fB\-cy\fP" options may be used to
choose the \fIc\fP parameter for the Julia set.
.IP "\fB\-verbose\fP"
Print a period on standard error for each reply received from a computation
server.
.IP "\fB\-statistics\fP"
Print performance statistics on standard error after finishing 
the calculations.
.IP "\fB\-ascii\fP"
Force ASCII format for the output file.
.PP
.SH BUGS
Bms normally generates binary
.I pgm
output,
but when the iteration count exceeds 256 it reverts to
the ASCII version of the file format.
This causes a large increase in file size and an even
larger increase in CPU usage.
.SH FILES
.if n .ta 2.5i
.if t .ta 2i
/usr/local/etc/mslaves	list of computation server hosts
.br
$HOME/.mslaves	per-user override of the above
.PP
.SH "SEE ALSO"
xms(1), mslaved(8), mslavedc(8), enslave(1), pgm(1), pnm(1)
.PP
.SH COPYRIGHT
.if n Copyright (C) 1990, 1991 Andreas Gustafsson (gson@niksula.hut.fi)
.if t Copyright \(co 1990, 1991 Andreas Gustafsson (gson@niksula.hut.fi)


ChangeLog
Wed Aug  6 17:23:33 1997  Andreas Gustafsson  <gson@guava.araneus.fi>

	* Wrapped version 0.07pl2

	* Switched from compress to gzip.

	* Updated contact information in README.

	* Commented out the X11R[23] support in xms/Imakefile
	because it misidentified X11R6 as X11R[23], and no-one
	uses X11R[23] anymore anyway.

	* 0.07pl1 fixed select() bug that slowed down bms to one packet
        per second.

Sat Sep  4 03:08:49 1993  Andreas Gustafsson  (gson@guano)

	* Wrapped version 0.07

	* Fixed two bugs in Ms.c which caused core dumps under
	Linux

Sun Aug 15 12:11:38 1993  Andreas Gustafsson  (gson@guano)

	* Added Linux FNDELAY hack
	
	* Added strdup() replacement (needed by pure 4.3BSD)

Sun Jun  6 04:27:00 1993  Andreas Gustafsson  (gson@guano)

	* Rewrote parts of the README file.

	* Gave up futile attempts to declare malloc() using system 
	include files.

	* Combined the various INSTALL files into a single file.

	* Fixed an use of freed memory and several memory leaks.

Sat May 22 03:04:12 1993  Andreas Gustafsson  (gson@guano)

	* Updated years in copyright messages.

Fri May  7 04:41:48 1993  Andreas Gustafsson  (gson@guano)

	* Version 0.06g ported to HP700-series and DEC Alpha
	machines.  The rubberband box can now be manipulated
	while the calculation is in progress.

	* Version 0.06h was put under RCS.  This necessitated a 
	restructuring of the code into a more traditional 
	directory tree with no links between directories.  There
	is now a lib/ directory for code shared between the
	different programs.  PPM support was added to bms.

	* Version 0.06f introduced experimental Henon fractal code
	in mslaved and associated generic parameter decoding code.

	* Version 0.06d1 changed to support a BitMapUnit of 32 in
	b/w operation.  Maybe it was a mistake to remove the
	8-bit code.

Thu May 14 19:58:16 1992  Andreas Gustafsson  (gson@psykoman.cs.hut.fi)

	* Lots of changes to support 24-bit visuals

Sat Feb  8 21:15:44 1992  Andreas Gustafsson  (gson at gunnar)

	* Added support for pipe operation in mslaved.c; introduced
	HAVE_SOCKETS and HAVE_SYSLOG compile-time flags

Wed Feb  5 16:57:41 1992  Andreas Gustafsson  (gson at spiff.cs.hut.fi)

	* Made the visual matching code	use the default visual for 
	black-and-white operation

	* Added colormap animation

Fri Dec  6 17:32:39 1991  Andreas Gustafsson  (gson at bulldozer.cs.hut.fi)

	* Allowed empty lines and # comments in .enslave

Sun Dec  1 16:18:26 1991  Andreas Gustafsson  (gson at punisher.cs.hut.fi)

	* New development version 0.06a.  Includes visual matching code in
	Mama.c, and new option handling in mslaved and mslavedc.

Sun Nov 24 12:38:12 1991  Andreas Gustafsson  (gson at hydroman.cs.hut.fi)

	* Wrapped version 0.06

	* Added option details to the synopsis in xms.man

	* Tested the new ms_draw() on monochrome and 8-bit colour
	Suns.

	* Completely rewrote the black-and-white special case display
	support in Ms.c:ms_draw(), so that it now constructs the image
	a byte at a time rather than a word at a time.  This should
	avoid any byteorder problems.

Tue Nov  5 17:23:38 1991  Andreas Gustafsson  (gson at bulldozer.cs.hut.fi)

	* Version number is now 0.05p.  Changed some key bindings to be 
	more consistent with the new "zoom by default" policy, and
	updated the manual page correspondingly.

Mon Sep 30 01:01:40 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Added XtNinput argument to shell widget; that ought to
	fix the keyboard focus problem with OpenWindows (not tested).

Sun Sep 22 12:07:27 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Edited README and mslaved/Makefile; now gcc is stronly recommended
	only for old CISCs.  Made floating point the default for MIPS in
	ms_real.h as suggested by broadley's benchmarks.

	* restored the menu entries for using the old window but made the
	popup alternative more defaultish (no, I don't run spell on
	ChangeLogs). 

	* added #include <sys/file.h> to ms_ipc.h (needed for FNDELAY in
	HP-UX) 

	* fixed some syntax errors in mslaved.c

	* added #include <netinet/in.h> in Ms.c to get htonl() and ntohs()
	macro definitions (needed for HP-UX)

Sat Sep 14 00:09:20 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* #ifdef'ed away the menu entries for reusing the old window

Sun Aug 11 14:42:55 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Added "inverse box zooming" as suggested by RMS.  Still
	untested.

Sat Aug 10 10:32:51 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Added #include <errno.h> in work.c

	* Wrapped 0.05j and mailed to broadley for beta testing

	* Renamed split.c to mspawn.c

	* Tested 0.05h on SPARCstation IPC.  Fixed some bugs in the
	noImakefile and documented problems with missing imake and
	Aw libs.  Version with these fixes called 0.05i.

Fri Jul 26 14:23:57 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Fixed some more typos and made some other changes in the man
	pages

	* Wrapped 0.05h	and gave it to ufa@niksula.hut.fi for beta 
	testing 

Sun Jul 21 12:36:22 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Wrote the bms batch mode client and renamed the X client to
	xms.  This also caused lots of changes to both source and 
	documentation of xms because of extensive code sharing between the
	two clients.  Mandelbrot-dependent but X-independent
	code has been separated into a new module, mspawn.c, leaving
	code that is both X- and Mandelbrot-independent in work.c.
	The timeout system has been completely rewritten.

	* MIPS fixed-point code has been tested on a DECStation.

Wed Jun 26 17:01:04 1991  Andreas Gustafsson  (gson at gunnar)

	* Separated the X-independent workload distribution stuff
	into a new module, work.c.  Renamed all X-independent files
	to lowercase names to reduce ugliness.  Added missing "2" in
	MIPS fracmult2 function name; need to retest MIPS code before
	distributing it.

Tue Jun 18 21:34:57 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Fixed several typos in the man page


bms/bms.c
/* bms.c - batch-mode MandelSpawn client main program */

/*  
    This file is part of MandelSpawn, a network Mandelbrot program.

    Copyright (C) 1990-1993 Andreas Gustafsson

    MandelSpawn is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License, version 1,
    as published by the Free Software Foundation.

    MandelSpawn is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License,
    version 1, along with this program; if not, write to the Free 
    Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <stdio.h>
#include <string.h>
#include <errno.h>
extern int errno;		/* at least Sony's errno.h misses this */
#include <math.h>		/* for atof() */

/*
  It is impossible to declare malloc() in a portable way.
  Be prepared to change these declarations. 
*/
void *malloc(), *realloc();
void free();

#include "datarep.h"
#include "io.h"
#include "work.h"
#include "mspawn.h"
#include "version.h"
#include "color.h"

typedef struct {
	unsigned char comps[3];
} pixel8;

#define TIMEOUT 15000		/* default computation server timeout in millisecs */

/*
  Number of pixels to print on each line in PGM and PPM files
  (watch out for line length limit of 70 characters) 
*/
#define PGM_NPL 8
#define PPM_NPL 4

char *me;			/* name of program */

/* various global mode flags */

int done = 0;			/* set when all chunks have arrived */
int verbose = 0;		/* verbosity level */
int statistics = 0;		/* print statistics if set */
int pgm_ascii = 0;		/* force ASCII graymap format */
int nooutput = 0;		/* generate no output (for testing only) */

/* structure for global data */

typedef struct bms_state {
	ms_state ms;
	union {
		unsigned char *bytes;
		unsigned short *shorts;
	} frame;
} bms_state;

/* report a fatal error and die */

void error(s)
char *s;
{
	fprintf(stderr, "%s: %s\n", me, s);
	exit(1);
}

/* handle input from a computation server by drawing in an in-core buffer */

void ms_draw(client, client_data, data) /*ARGSUSED*/ char *client;
char *client_data;
char *data;
{
	ms_client_info *the_info = (ms_client_info *) client_data;
	bms_state *b = (bms_state *) client;
	unsigned int x, y, width, height;
	unsigned int i, j;
	if (verbose) {
		putc('.', stderr);
		fflush(stderr);
	}
	x = the_info->s.x;
	y = the_info->s.y;
	width = the_info->s.width;
	height = the_info->s.height;

	switch (b->ms.bytes_per_count) {
	case 1:
		{
			unsigned char *datap = (unsigned char *)data;
			unsigned char *bufp_left = (unsigned char *)
			    b->frame.bytes + b->ms.width * y + x;
			for (j = 0; j < height; j++) {
				unsigned char *bufp = bufp_left;
				for (i = 0; i < width; i++) {
					*bufp++ = *datap++;
				}
				bufp_left += b->ms.width;
			}
		}
		break;
	case 2:
		{
			unsigned short *datap = (unsigned short *)data;
			unsigned short *bufp_left = (unsigned short *)
			    b->frame.shorts + b->ms.width * y + x;
			for (j = 0; j < height; j++) {
				unsigned short *bufp = bufp_left;
				for (i = 0; i < width; i++) {
					*bufp++ = *datap++;
				}
				bufp_left += b->ms.width;
			}
		}
		break;
	default:
		error("bad iteration count size");
		break;
	}
}

/*
  provide error reporting services to the workforce and
  I/O modules 
*/

void wf_error(s)
char *s;
{
	error(s);
}

void wf_warn(s)
char *s;
{
	fprintf(stderr, "warning: %s\n", s);
}

void io_error(s)
char *s;
{
	error(s);
}

void io_warn(s)
char *s;
{
	fprintf(stderr, "warning: %s\n", s);
}

/* unique values to identify options */

enum bms_opt { opt_height, opt_width, opt_x, opt_y, opt_range,
	opt_julia, opt_colours, opt_cx, opt_cy,
	opt_chunk_width, opt_chunk_height, opt_verbose,
	opt_nooutput, opt_statistics, opt_version, opt_ascii,
	opt_colour, opt_spectrum
};

struct option {
	enum bms_opt id;
	char *name;
	int has_arg;
};

struct option options[] = { {opt_height, "height", 1},
{opt_width, "width", 1},
{opt_x, "x", 1},
{opt_y, "y", 1},
{opt_range, "range", 1},
{opt_julia, "julia", 0},
{opt_colours, "colours", 1},
{opt_colours, "colors", 1},
{opt_colours, "iterations", 1},
{opt_cx, "cx", 1},
{opt_cy, "cy", 1},
{opt_chunk_width, "chunk_width", 1},
{opt_chunk_height, "chunk_height", 1},
{opt_verbose, "verbose", 0},
{opt_nooutput, "nooutput", 0},
{opt_statistics, "statistics", 0},
{opt_version, "version", 0},
{opt_ascii, "ascii", 0},
{opt_colour, "colour", 0},
{opt_colour, "color", 0},
{opt_spectrum, "spectrum", 1}
};

cmap_error(msg)
char *msg;
{
	error(msg);
}

int main(argc, argv)
int argc;
char **argv;
{
	bms_state bms;
	wf_state *workforce;
	int i, j;
	char *optarg;

	/* colour stuff */
	int colour_mode = 0;
	char *spectrum =
	    "blue-aquamarine-cyan-medium sea green-forest green-lime green-\
yellow green-yellow-coral-pink-black";
	color_t *colors = 0;
	pixel8 *pixels = 0;

	me = argv[0];

	/* defaults */
	bms.ms.width = 64;
	bms.ms.height = 24;
	bms.ms.center_x = -0.5;
	bms.ms.center_y = 0.0;
	bms.ms.xrange = 4.0;
	bms.ms.julia = 0;
	/* bms.ms.c_x... */
	bms.ms.job.iteration_limit = 250;

	bms.ms.chunk_height = bms.ms.chunk_width = 32;

	for (i = 1; i < argc; i++) {
		char *arg = argv[i];
		if (*arg == '-')
			for (j = 0; j < sizeof(options) / sizeof(options[0]);
			     j++) {
				if (strcmp(arg + 1, options[j].name) == 0) {
					if (options[j].has_arg) {
						if (argc < i)
							error
							    ("option requires argument");
						optarg = argv[++i];
					}
					switch (options[j].id) {
					case opt_width:
						bms.ms.width = atoi(optarg);
						break;
					case opt_height:
						bms.ms.height = atoi(optarg);
						break;
					case opt_x:
						bms.ms.center_x = atof(optarg);
						break;
					case opt_y:
						bms.ms.center_y = atof(optarg);
						break;
					case opt_range:
						bms.ms.xrange = atof(optarg);
						break;
					case opt_julia:
						bms.ms.julia = 1;
						break;
					case opt_colours:
						bms.ms.job.iteration_limit =
						    atoi(optarg);
						break;
					case opt_cx:
						bms.ms.c_x = atof(optarg);
						break;
					case opt_cy:
						bms.ms.c_y = atof(optarg);
						break;
					case opt_chunk_width:
						bms.ms.chunk_width =
						    atoi(optarg);
						break;
					case opt_chunk_height:
						bms.ms.chunk_height =
						    atoi(optarg);
						break;
					case opt_verbose:
						verbose++;
						break;
					case opt_nooutput:
						nooutput++;
						break;
					case opt_statistics:
						statistics++;
						break;
					case opt_version:
						printf("bms version %s\n",
						       ms_version);
						exit(0);
						break;
					case opt_ascii:
						pgm_ascii++;
						break;
					case opt_colour:
						colour_mode++;
						break;
					case opt_spectrum:
						spectrum = optarg;
						break;
					default:
						error
						    ("internal option procesing error");
					}
					goto nextopt;
				}	/* if !strcmp */
			}	/* for j */
		error("unrecognized option");
 nextopt:	;
	}			/* for i */

	if (colour_mode) {
		int i;
		int ncolours = bms.ms.job.iteration_limit;

		colors = (color_t *) malloc(ncolours * sizeof(color_t));
		pixels = (pixel8 *) malloc(ncolours * sizeof(pixel8));
		parse_spectrum(spectrum, ncolours, colors);

		for (i = 0; i < ncolours; i++) {
			int comp;
			for (comp = 0; comp < 3; comp++)
				pixels[i].comps[comp] =
				    (int)(0xFF * colors[i].comps[comp] + 0.5);
		}
	}

	bms.ms.bytes_per_count = (bms.ms.job.iteration_limit > 256 ? 2 : 1);

	ms_init(&bms.ms, (char *)&bms,
		workforce = wf_init(TIMEOUT, IO_MUX_NONE, IO_MUX_SELECT));

	ms_calculate_job_parameters(&bms.ms, &bms.ms.job);

	/* allocate memory for frame buffer */
	switch (bms.ms.bytes_per_count) {
	case 1:
		bms.frame.bytes = (unsigned char *)
		    malloc(bms.ms.width * bms.ms.height *
			   sizeof(unsigned char));
		break;
	case 2:
		bms.frame.shorts = (unsigned short *)
		    malloc(bms.ms.width * bms.ms.height *
			   sizeof(unsigned short));
		break;
	default:
		error("bad iteration count size");
		break;
	}

	ms_dispatch_rect(&bms.ms, (char *)&bms, 0, 0, bms.ms.width,
			 bms.ms.height);

	ms_main(&bms.ms);

	/* we have received all the replies; write out the finished picture */
	if (!nooutput) {
		if (verbose) {
			putc('\n', stderr);
		}

		/* force ASCII mode if PGM maxval > 255 */
		if (!colour_mode && bms.ms.bytes_per_count > 1)
			pgm_ascii = 1;

		printf("P%d\n%d %d\n%d\n",
		       2 + colour_mode + 3 * (!pgm_ascii),
		       bms.ms.width, bms.ms.height,
		       colour_mode ? 0xFF : bms.ms.job.iteration_limit - 1);

		if (!colour_mode) {
			if (pgm_ascii) {	/* ASCII pgm format */
				unsigned i, j;
				for (j = 0; j < bms.ms.height; j++) {
					for (i = 0; i < bms.ms.width; i++)
						printf("%d%c",
						       bms.ms.bytes_per_count ==
						       2 ? bms.frame.shorts[j *
									    bms.
									    ms.
									    width
									    +
									    i] :
						       bms.frame.bytes[j *
								       bms.ms.
								       width +
								       i],
						       i % PGM_NPL ==
						       (PGM_NPL -
							1) ? '\n' : ' ');
					if (i % PGM_NPL)
						putchar('\n');
				}
			} else {	/* binary pgm format */
				unsigned i, j;
#ifdef PARANOIA
				for (j = 0; j < bms.ms.height; j++)
					for (i = 0; i < bms.ms.width; i++)
						putchar(bms.frame.
							bytes[j * bms.ms.width +
							      i]);
#else
				/* we happen to use the same internal representation for the image */
				/* as the data part of a binary pgm file, so we can dump it as such */
				fwrite(bms.frame.bytes, sizeof(char),
				       bms.ms.width * bms.ms.height, stdout);
#endif
			}
		} else {	/* colour */

			if (pgm_ascii) {	/* ASCII ppm format */
				unsigned i, j;

				for (j = 0; j < bms.ms.height; j++) {
					for (i = 0; i < bms.ms.width; i++) {
						pixel8 *pixel =
						    &pixels[bms.ms.
							    bytes_per_count ==
							    2 ? bms.frame.
							    shorts[j *
								   bms.ms.
								   width +
								   i] : bms.
							    frame.bytes[j *
									bms.ms.
									width +
									i]
						    ];
						printf("%d %d %d%s",
						       pixel->comps[RED_COMP],
						       pixel->comps[GREEN_COMP],
						       pixel->comps[BLUE_COMP],
						       i % PPM_NPL ==
						       (PPM_NPL -
							1) ? "\n" : "  ");
					}
					if (i % PPM_NPL)
						putchar('\n');
				}
			} else {	/* binary ppm format */
				unsigned i, j;

				for (j = 0; j < bms.ms.height; j++)
					for (i = 0; i < bms.ms.width; i++) {
						pixel8 *pixel =
						    &pixels[bms.ms.
							    bytes_per_count ==
							    2 ? bms.frame.
							    shorts[j *
								   bms.ms.
								   width +
								   i] : bms.
							    frame.bytes[j *
									bms.ms.
									width +
									i]
						    ];
						putchar(pixel->comps[RED_COMP]);
						putchar(pixel->
							comps[GREEN_COMP]);
						putchar(pixel->
							comps[BLUE_COMP]);
					}
			}
		}
	}

	if (statistics) {
		wf_print_stats(workforce, stderr);
	}
	exit(0);
}
